docs: migrate what is angular with cup of coffee (#51378)
angular/angular: aio/content/guide/what-is-angular.md

As a platform, Angular includes:

- A component-based framework for building scalable web applications
- A collection of well-integrated libraries that cover a wide variety of features, including routing, forms management, client-server communication, and more
- A suite of developer tools to help you develop, build, test, and update your code

With Angular, you're taking advantage of a platform that can scale from single-developer projects to enterprise-level applications. Best of all, the Angular ecosystem consists of a diverse group of over 1.7 million developers, library authors, and content creators.

<a id="essentials"></a>

## Prerequisites

Like most modern frameworks, Angular expects you to be familiar with HTML, CSS and JavaScript. In addition, it’s recommended to have familiarity with the following concepts and tools:

<a id="concepts"></a>

### Concepts

- [JavaScript Classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)
- [TypeScript fundamentals](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html)
- [TypeScript Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html)

<a id="tools"></a>

### Tools

> - **TypeScript** - This is shipped by default with every Angular app to provide improved tooling and enhanced maintainability for a better developer experience

Reviewer: ```suggestion
- **TypeScript** - This is shipped by default with every Angular app to provide improved tooling and enhanced maintainability for a better developer experience.
```

- **Command Line Interface (CLI)** - Angular uses a compiler in order to abstract tooling complexity and optimize your code so you can focus on building your app.

<a id="components"></a>

## Components

Components are the fundamental building block for creating applications in Angular. By leveraging component architecture, Angular aims to provide structure for organizing your project into manageable, well organized parts with clear responsibilities so that your code is maintainable and scalable.

> An Angular component can be identified by the `component` suffix (e.g., `my-custom-component.component.ts` and has the following:

Reviewer: ```suggestion
An Angular component can be identified by the `component` suffix (e.g., `my-custom-button.component.ts` and has the following:
```

Wouldn't it be better if it was sth more meaningful, as having `component` twice may confuse devs.

Author: This is true. I was trying to make it more generic, so maybe `my-custom-name` would be better? What do you think?

Author: I thought about your suggestion and saw that this is being repeated in `service` and `directive`, so currently going with the `my-custom-name` to keep it consistent. 

Let me know if you come up with a better suggestion though!

> - A TypeScript decorator to define configuration options for things like:

Reviewer: ```suggestion
- A decorator to define configuration options for things like:
```

Angular decorators are removed at build time so they are not 100% TypeScript decorators.

Here is a simplified example of a TodoListItem component.

```ts
// 📄 todo-list-item.component.ts
@Component({
  standalone: true,
  selector: 'todo-list-item',
  template: ` <li>(TODO) Read cup of coffee introduction</li> `,
  styles: ['li { color: papayawhip; }'],
})
export class TodoListItem {
  /* Component behavior is defined in here */
}
```

<a id="behavior"></a>

### Behavior

Now that we have the basic structure for the component, let’s learn more about how you define the component’s behavior.

#### State

When defining data that you want the component to manage, this can be accomplished by declaring it by defining [class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields).

In the example of a `todo-list-item.component.ts`, there are two properties we want to track: `taskTitle` and `isComplete`. Using the [class field syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields), they can be defined as follows:

```ts
// 📄 todo-list-item.component.ts
@Component({ ... })
export class TodoList {
  taskTitle = '';
  isComplete = false;
}
```

#### Methods

> You can define functions for a component by declaring functions within the component class.

Reviewer: ```suggestion
You can define functions for a component by declaring methods within the component class.
```

```ts
// 📄 todo-list-item.component.ts
@Component({ ... })
export class TodoList {
  taskTitle = '';
  isComplete = false;

  updateTitle(newTitle: string) {
    this.taskTitle = newTitle;
  }

  completeTask() {
    this.isComplete = true;
  }
}
```

<a id="templates"></a>

### Templates

Every component has an HTML template that defines what that component renders to the DOM.

HTML templates can be defined as an inline template within the TypeScript class, or in separate files with the `templateUrl` property. To learn more, check out [the docs on defining component templates](https://angular.io/guide/component-overview#defining-a-components-template).

Within this document, the examples will use inline templates for more concise code snippets.

#### Rendering Dynamic Data

When you need to display dynamic content in your template, Angular uses the double curly brace syntax in order to distinguish between static and dynamic content.

```ts
@Component({
  template: ` <p>Title: {{ taskTitle }}</p> `,
})
export class TodoListItem {
  taskTitle = 'Read cup of coffee';
}
```

This is how it renders to the page.

```html
<p>Title: Read cup of coffee</p>
```

> This syntax declares a **binding** between the dynamic data property inside of the HTML. As a result, whenever the data changes, Angular will automatically update the DOM reflecting the new value of the property.

Reviewer: ```suggestion
This syntax declares an **interpolation** between the dynamic data property inside of the HTML. As a result, whenever the data changes, Angular will automatically update the DOM reflecting the new value of the property.
```

In Angular {{ }} syntax is mostly referred as interpolation: https://angular.io/guide/interpolation

> #### Dynamic Attributes

Reviewer: ```suggestion
#### Dynamic Properties and Attributes
```

In Angular we distinct attributes from properties when we bind to them: 
- https://angular.io/guide/property-binding
- https://angular.io/guide/attribute-binding

When you need to dynamically set the value of attributes in an HTML element, the target property is wrapped in square brackets. This binds the attribute with the desired dynamic data by informing Angular that the declared value should be interpreted as a JavaScript-like statement ([with some Angular enhancements](https://angular.io/guide/understanding-template-expr-overview)) instead of a plain string.

```html
<button [disabled]="hasPendingChanges"></button>
```

In this example, the disabled property is tied to the `hasPendingChanges` variable that Angular would expect to find within the component’s state.

#### Event Handling

You can bind event listeners by specifying the event name in parenthesis and invoking a method on the right-hand-side of the equals sign:

```html
<button (click)="saveChanges()">Save Changes</button>
```

> If you need to pass the event object to your event listener, Angular provides an implicit `$event` variable that can be used inside the function call:

Reviewer: ```suggestion
If you need to pass the event object to your event listener, Angular provides an implicit `$event` variable that can be used inside the function call:
```

We may also mention that the type of the $event is based on the emitter of the event. So it's totally typechecked.

Author: That's good to know! That detail is probably better explained in an in-depth section on this topic so people can learn about what makes it awesome. Appreciate the comment!

```html
<button (click)="saveChanges($event)">Save Changes</button>
```

<a id="styles"></a>

### Styles

> When you need to style a component, there are two optional properties that you can configure inside of the `@Component` decorator.

Author: If it's not currently used a lot in the community, we can leave it out for now since this guide was written from the perspective of giving someone a high level overview of how Angular works.

Thanks for letting me know though! Something we'll want to document in an in-depth guide later for sure.

Similar to component templates, you can manage a component's styles in the same file as the TypeScript class, or in separate files with the `styleUrls` properties.

Components can optionally include a list of CSS styles that apply to that component's DOM:

```ts
@Component({
  selector: 'profile-pic',
  template: `<img src="profile-photo.jpg" alt="Your profile photo" />`,
  styles: [
    `
      img {
        border-radius: 50%;
      }
    `,
  ],
})
export class ProfilePic {
  /* Your code goes here */
}
```

By default, a component's style will only apply to elements in that component's template in order to limit the side effects.

To learn more, check out [the docs on component styling](https://angular.io/guide/component-styles).

<a id="directives"></a>

## Directives

> When building applications, developers often need to extend on the behavior of an HTML element. Examples of this include: displaying content based on a certain condition, rendering a list of items based on application data, changing the styles on an element based on user interaction, etc.

Reviewer: ```suggestion
When building applications, developers often need to extend on the behavior of an HTML element or Angular directives/components. Examples of this include: displaying content based on a certain condition, rendering a list of items based on application data, changing the styles on an element based on user interaction, etc.
```

We can apply directives in html, other directives or components and not just HTML elements.

<a id="services"></a>

## Services

When you need to share logic between components, Angular allows you to create a “service” which allows you to inject code into components while managing it from a single source of truth.

Angular services can be identified by the `service` suffix (e.g., `my-custom-service.service.ts`).

Similar to defining a component, services are comprised of the following:

- A TypeScript decorator to define configuration options for things like:
  - [`providedIn`](https://angular.io/api/core/Injectable#providedIn) - This allows you to define what parts of the application can access the service. For example, ‘root’ will allow a service to be accessed anywhere within the application.
- A TypeScript class that defines the desired code that will be accessible when the service is injected

Here is an example of a `Calculator` service.

```ts
import { Injectable } from '@angular/core';

@Injectable({
  providedIn: 'root',
})

> class Calculator {

Reviewer: ```suggestion
class CalculatorService {
```
Just like we add Component or Directive or Pipe to the end of the class we do the same for Services too.

Author: Good to know!

  add(x: number, y: number) {
    return x + y;
  }
}
```

If we wanted to call the service in a Receipt component for example, here’s what it might look like:

```ts
import { Component } from '@angular/core';

> import { Calculator } from './calculator.service';

Reviewer: ```suggestion
import { CalculatorService } from './calculator.service';
```

@Component({
  selector: 'app-receipt’,
  template: `<p>The total is {{ totalCost }}</h1>`,
})
export class Receipt {

>   private calculatorService = inject(Calculator);

Reviewer: ```suggestion
  private calculatorService = inject(CalculatorService);
```

  totalCost = this.calculatorService.add(50, 25);
}
```

> In this example, the `Calculator` is being used by calling the Angular function `inject` and passing in the service to it.

Reviewer: ```suggestion
In this example, the `CalculatorService` is being used by calling the Angular function `inject` and passing in the service to it.
```

<a id="organization"></a>

## Organization

Standalone components are a new organizational pattern that were introduced in Angular v15 and is the recommended place to start. In contrast to [NgModules](https://angular.io/guide/ngmodules), it allows developers to organize code and manage dependencies through components rather than feature modules.

> For example, in the traditional NgModule pattern, you would need to create a TodoApp module and manage all of its dependencies through this module.

Reviewer: ```suggestion
For example, in the traditional NgModule pattern, you would need to create a TodoModule and manage all of its dependencies through this module.
```

To be the same as the example

})
export class TodoModule {}
```

However, you can now achieve something similar with a standalone component without the need for a module file:

```ts (Todo.component.ts)
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TodoList } from '../todo/todo-list.component';

@Component({
  standalone: true,
  selector: 'todo-app',
  imports: [FormsModule, TodoList],
  template: ` ... <todo-list [tasks]="taskList"></todo-list> `,
})
export class PhotoGalleryComponent {
  // component logic
}
```

While most of this should be familiar (from the Components section), two things that are unique to this new pattern are the `standalone` flag and the `imports` key.

- `standalone` - When provided the value `true`, this tells Angular that the component does not need to be declared in an NgModule
- `imports` - Allows developers to declare what dependencies will be used in the component

In other words, rather than having to define a specific context in which code should be organized, developers are able to specify the dependencies directly within the component context itself.

<a id="cli"></a>

## Command Line Interface (CLI)

> The Angular CLI is the recommended way to develop Angular applications and can make some tasks trouble-tree.

Reviewer: ```suggestion
The Angular CLI is the recommended way to develop Angular applications and can make some tasks trouble-free.
```

#### State

When defining data that you want the component to manage, this can be accomplished by declaring it by defining [class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields).

In the example of a `todo-list-item.component.ts`, there are two properties we want to track: `taskTitle` and `isComplete`. Using the [class field syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields), they can be defined as follows:

```ts
// 📄 todo-list-item.component.ts
@Component({ ... })
export class TodoList {
  taskTitle = '';
  isComplete = false;
}
```

#### Methods

You can define functions for a component by declaring methods within the component class.

```ts
// 📄 todo-list-item.component.ts
@Component({ ... })
export class TodoList {
  taskTitle = '';
  isComplete = false;

  updateTitle(newTitle: string) {
    this.taskTitle = newTitle;
  }

  completeTask() {
    this.isComplete = true;
  }
}
```

<a id="templates"></a>

### Templates

Every component has an HTML template that defines what that component renders to the DOM.

> HTML templates can be defined as an inline template within the TypeScript class, or in separate files with the `templateUrl` property. To learn more, check out [the docs on defining component templates](https://angular.io/guide/component-overview#defining-a-components-template).

Reviewer 2: And this one !

Author: Thanks for catching these! I've updated the PR to migrate them to relative links!

</ul>
```

### Custom directives

While built-in directives help to solve common problems that developers encounter, there are situations where developers require custom behavior that’s specific to their application. In these cases, Angular provides a way for you to create custom directives.

Custom Angular directives can be identified by the `directive` suffix (e.g., `my-custom-name.directive.ts`).

Similar to defining a component, directives are comprised of the following:

- A TypeScript decorator to define configuration options for things like:
  - A selector that defines the tag name is when the component is called
- A TypeScript class that defines the extended behavior the directive will add to the respective HTML element.

For example, here’s what a custom directive for highlighting an element:

```ts
@Directive({
  selector: '[appHighlight]',
})
export class HighlightDirective {
  private el = inject(ElementRef);
  constructor() {
    this.el.nativeElement.style.backgroundColor = 'yellow';
  }
}
```

To apply this to an element, the directive is called by adding it as an attribute.

```html
<p appHighlight>Look at me!</p>
```

> Directives can also leverage user events, take input for additional customization, but this goes beyond the scope of this article. To learn more, check out [the docs on creating custom directives](https://angular.io/guide/attribute-directives).

Reviewer 2: This one also !
